Docker ubuntu 18.04 support and static execs

nilsdeppe opened the pull request:
## Proposed changes

This adds most of the remaining code needed to CD the CCE executables.

The container is already push to the Dockerhub sxscollaboration image. I previously tested on my fork.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Support Ubuntu 18.04 container

This is needed for deploying static libraries and executables for people to use
since we need a fairly old version of glibc.

nilsdeppe committed:
Use static libcrypto and libz with static HDF5

nilsdeppe committed:
Improve static fortran lib support

nilsdeppe committed:
Add deploy_static_execs_and_libs image to Dockerfile

This image can be used for deploy static library builds of anything that we want
usable by other groups.

nilsdeppe committed:
Update DemoContainer action versions

nilsdeppe committed:
Switch to softprops/action-gh-release for releases

nilsdeppe committed:
Clarify what and how to push Docker containers to Dockerhub

Waiting for these to build in CI can be slow, so being able to build them
locally and push them is useful.

nilsdeppe committed:
Build dev1804 container in CI when needed

nilsdeppe committed:
Add DeployStaticExecutables workflow

nilsdeppe commented:
I've rebased this and set up auto deployment. I'm doing a final test on my fork to make sure it's working correctly, but I expect anything wrong would be a small typo in the last step of testing the CCE output is correct. This now bundles the `CharacteristicExtract` and `ReduceCceWoldtube` executables along with a `BondiSachsR0200.h5` worldtube file, expected output file, and a python script that checks that things went okay. The last 3 parts are housed in our Google Drive since that was easiest to get working quick.

edit:

The release on my fork is running at https://github.com/nilsdeppe/spectre/actions/runs/9374720262/job/25811397363

nilsdeppe commented:
This successfully deployed the executables on my fork. https://github.com/nilsdeppe/spectre/releases/tag/v2024.06.04.013

nilsdeppe committed:
Install h5py from apt in container builds

This speeds up cross-compilation considerably.

nilsdeppe committed:
Document the release token in Tests.yaml

I needed to know this when testing releases on my fork and it would've been nice
to have this info.

nilsvu commented on .github/workflows/BuildDockerContainer.yaml:
```diff
@@ -73,6 +73,37 @@ jobs:
+          target: dev
+          tags: ${{ inputs.image_name }}:dev1804
+          platforms: linux/amd64
+          build-args:
```
I think this needs a `|`: `build-args: |`

nilsvu commented on .github/workflows/DeployStaticExecutables.yaml:
```diff
@@ -0,0 +1,106 @@
+name: Deploy static executables
+
+on:
+  release:
+    types: [published]
```
Could you switch this to:

```yaml
  workflow_call:
    inputs:
      tag_name: ... # same as below
    secrets:
      DOCKERHUB_USERNAME:
        required: true
      DOCKERHUB_TOKEN:
        required: true
```

Then at the bottom of `PostRelease.yaml` add:

```yaml
  static_execs:
    uses: ./.github/workflows/DeployStaticExecutables.yaml
    secrets: inherit
    with:
      tag_name: ${{ inputs.tag_name || github.event.release.tag_name }}
```

Then this runs as a job in the `PostRelease` workflow together with building the other containers.

nilsdeppe replied:
Sorry, do I need the secrets? This doesn't push to DockerHub so I don't think I need the username or token. This only pushes assets to the release. I had a lot of troubles with the `workflow_call` so I'd rather keep this as-is. Essentially, it wasn't clear to me what was going on and going back and testing it again takes a decent amount of time that I don't really have

nilsvu replied:
No you're right you don't need the secrets

nilsvu commented on .github/workflows/DeployStaticExecutables.yaml:
```diff
@@ -0,0 +1,106 @@
+    runs-on: ubuntu-latest
+    # Running in a protected environment that provides the necessary secrets
+    environment: release
+    env:
+      RELEASE_TAG_NAME: ${{ inputs.tag_name || github.event.release.tag_name }}
```
Don't need this anymore when you make the change above. You can use `inputs.tag_name` directly.

nilsvu commented on .github/workflows/DeployStaticExecutables.yaml:
```diff
@@ -0,0 +1,106 @@
+    name: Deploy static executables and libraries
+    runs-on: ubuntu-latest
+    # Running in a protected environment that provides the necessary secrets
+    environment: release
```
Also use `deploy-containers` environment? At least that's what DemoContainers.yaml does.

nilsdeppe replied:
The deploy doesn't have access to the release token.

nilsvu replied:
You don't need the release token I think. But `deploy-containers` isn't the right environment either because you don't push a container. Question is, do you want to get prompted again for approval before the static execs get uploaded? Then keep the `release` env. Otherwise, I think you can just delete it.

nilsdeppe replied:
I don't think we need to get prompted before upload? There is a check that the CCE run is consistent with previous results, so the action should fail if something went wrong. I guess then I'll just not specify an env. I'll do this and push. If I misunderstood just let me know and I'll fix it :)

nilsvu commented on .github/workflows/DeployStaticExecutables.yaml:
```diff
@@ -0,0 +1,106 @@
+          platforms: linux/amd64
+          load: true
+          build-args:
+            UBUNTU_VERSION=18.04
```
Add `PARALLEL_MAKE_ARG=-j4`?

nilsvu commented on .github/workflows/DeployStaticExecutables.yaml:
```diff
@@ -0,0 +1,106 @@
+          ./CceExecutables/CharacteristicExtract
+          --input-file ./CceExecutables/CharacteristicExtract.yaml
+
+          python ./CheckCceOutput.py
```
Don't you have to point this script to the data?

nilsdeppe replied:
No, the names are hard-coded :shrug:

nilsvu commented on .github/workflows/DeployStaticExecutables.yaml:
```diff
@@ -0,0 +1,106 @@
+            docker cp static-execs:/work/spectre/build/bin/$i ./CceExecutables;
+          done
+
+          # Download extra data to include for CCE
```
(optional) I'd split this into multiple `run` steps so it's easier to track on CI (where time is spent, what fails, etc). No change to the script, just insert:

```yaml
- name: Download extra data to include for CCE
  run: >
    gdown ...
```

Also helps inserting steps like testing the exporter lib.

nilsvu commented on .github/workflows/DeployStaticExecutables.yaml:
```diff
@@ -0,0 +1,106 @@
+          mv CheckCceOutput.py ./CceExecutables/
+
+          tar cJf CceExecutables.tar.xz CceExecutables
+      - name: Create release on GitHub
```
"Upload to release"

nilsvu commented on .github/workflows/DeployStaticExecutables.yaml:
```diff
@@ -0,0 +1,106 @@
+      - name: Create release on GitHub
+        uses: softprops/action-gh-release@v2
+        with:
+          token: ${{ secrets.GH_TOKEN_RELEASE }}
```
You can delete this line (the PAT is only needed for git pushes etc, not to just upload a file, I believe).

nilsdeppe replied:
I would be a bit surprised by that and don't really have time to test that. Is there a major drawback to having it?

nilsdeppe committed:
Switch to dev container in CI

nilsdeppe committed:
Only build charm++ instead of the LIBS in the container

We only need the LIBS target if we do load-balancing, which we are switching to
nodegroups for anyway.

nilsvu commented on containers/Dockerfile.buildenv:
```diff
@@ -31,18 +60,49 @@
 #                 -f ./containers/Dockerfile.buildenv .
 #
-# 2. `ci` (this should use the `dev` image you just built for most of it)
+#   # To cross compile for arm64 and amd64 use
```
delete this line?

nilsdeppe replied:
Reworded to `Cross compiling for AMD64 and ARM64:` to mark it as a "section"

nilsvu commented on containers/Dockerfile.buildenv:
```diff
@@ -110,12 +220,31 @@ RUN apt-get update -y \
 # - We use python-is-python3 because on Ubuntu 20.04 /usr/bin/python was removed
 #   to aid in tracking down anything that depends on python 2. However, many
 #   scripts use `/usr/bin/env python` to find python so restore it.
+# - We install h5py from apt since that's much faster when cross compiling than
```
do you mean "when building the old Ubuntu 18.04 container"? This isn't related to cross-compiling, right? Is the issue that for the old ubuntu no precompiled h5py is available from pip, so it tries to build wheels?
Edit: wait this is in the new container. It should not build wheels (compile h5py from source). But I remember seeing this problem too. Try to delete the h5py version constraint from `support/Python/requirements.txt`.

nilsdeppe replied:
Resolved offline: now does `pip3 --no-cache-dir install --only-binary=h5py...`

nilsvu commented on containers/Dockerfile.buildenv:
```diff
@@ -124,24 +253,34 @@ RUN apt-get update -y \
 # Install build dependencies for the host system
 RUN apt-get update -y \
-    && apt-get install -y cmake clang lld wget
+    && apt-get install -y cmake clang lld flex bison wget git
```
Can you drop these changes to the `xbuild` stage?

nilsdeppe replied:
No, these were necessary. I need bison and flex on the host arch, not the target arch.

nilsvu commented on containers/Dockerfile.buildenv:
```diff
@@ -163,13 +302,28 @@ RUN wget https://github.com/doxygen/doxygen/archive/Release_1_9_3.tar.gz -O doxy
+    fi
 
 
+# WARNING: We are now back in the spectre container but we have 2, one for each
```
The docker build doesn't know anything about the other architecture it's building for. I'd delete the "we have 2" part of this comment.

nilsdeppe replied:
Sorry, how do we not have two? Below it says `base-amd64` and `base-arm64`

nilsvu replied:
From the two stages `base-amd64` and `base-arm64` only the arch we're actually building is used. The line `FROM base-${TARGETARCH} AS dev` selects it. This is effectively an `if` in the Dockerfile (it's annoying that the Dockerfile syntax doesn't make that easier).

nilsdeppe replied:
Ah okay. What do you think of the wording:

We are now back in the spectre container. We create two base images for the 2 possible target architectures and then select the one we need as the `dev` container below. This is essentially an `if` in the Dockerfile.

nilsvu replied:
Sounds good 👍

nilsdeppe committed:
Remove the ci container from the docker file

nilsdeppe reviewed (commented):
I've pushed some fixups. Thanks for the review!

nilsvu commented:
You can squash the fixups 👍

nilsdeppe commented:
I've squashed, rebased, and squashed in a few of the other changes discussed. I think the bigger thing I'm hesitant to do is the `workflow_call` part. I'd tried something like that previous and must've done something wrong because it didn't work at all as expected.

nilsvu commented on .github/workflows/DeployStaticExecutables.yaml:
```diff
@@ -0,0 +1,92 @@
+      # of Ubuntu so that the glibc is compatible with most Linux distros.
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v3
+      - name: Build and push deploy container
```
"Build static executables"

nilsvu commented on containers/Dockerfile.buildenv:
```diff
@@ -110,12 +203,30 @@ RUN apt-get update -y \
+    && rm -rf ./Python-3.10.1.tgz ./Python-3.10.1 \
+    && python3.10 -m pip install --upgrade pip \
+    && pip3.10 --no-cache-dir install pybind11~=2.6.1 \
+    && pip3.10 --no-cache-dir install -r requirements.txt \
```
Also add `--only-binary` here?

nilsvu commented on containers/Dockerfile.buildenv:
```diff
@@ -124,24 +235,34 @@ RUN apt-get update -y \
 # Install build dependencies for the host system
 RUN apt-get update -y \
-    && apt-get install -y cmake clang lld wget
+    && apt-get install -y cmake clang lld flex bison wget git
```
Drop the changes until line 303? No changes to the xbuild stage should be necessary

nilsdeppe replied:
No, these were necessary to get cross compiling to work.

nilsvu reviewed (changes requested):
Ok the `on: release` is fine too

nilsdeppe commented:
I squashed the `--only-binary` change
